rechunker/algorithm.py

```diff
@@ -125,40 +168,105 @@ def rechunking_plan(
             f"Target chunk memory ({target_chunk_mem}) exceeds max_mem ({max_mem})"
         )
 
+    if max_mem < min_mem:  # basic sanity test
+        raise ValueError(
+            "max_mem ({max_mem}) cannot be smaller than min_mem ({min_mem})"
+        )
+
     if consolidate_writes:
         write_chunks = consolidate_chunks(shape, target_chunks, itemsize, max_mem)
     else:
         write_chunks = tuple(target_chunks)
 
     if consolidate_reads:
-        read_chunk_limits: List[Optional[int]]
-        read_chunk_limits = []  #
-        for n_ax, (sc, wc) in enumerate(zip(source_chunks, write_chunks)):
-            read_chunk_lim: Optional[int]
+        read_chunk_limits: List[Optional[int]] = []
+        for sc, wc in zip(source_chunks, write_chunks):
+            limit: Optional[int]
             if wc > sc:
                 # consolidate reads over this axis, up to the write chunk size
-                read_chunk_lim = wc
+                limit = wc
             else:
                 # don't consolidate reads over this axis
-                read_chunk_lim = None
-            read_chunk_limits.append(read_chunk_lim)
+                limit = None
+            read_chunk_limits.append(limit)
 
         read_chunks = consolidate_chunks(
             shape, source_chunks, itemsize, max_mem, read_chunk_limits
         )
     else:
         read_chunks = tuple(source_chunks)
 
-    # Intermediate chunks  are the smallest possible chunks which fit
-    # into both read_chunks and write_chunks.
-    # Example:
-    #   read_chunks:            (20, 5)
-    #   target_chunks:          (4, 25)
-    #   intermediate_chunks:    (4, 5)
-    # We don't need to check their memory usage: they are guaranteed to be smaller
-    # than both read and write chunks.
-    intermediate_chunks = [
-        min(c_read, c_target) for c_read, c_target in zip(read_chunks, write_chunks)
-    ]
+    # increase the number of stages until min_mem is exceeded
+    for stage_count in range(MAX_STAGES):
+
+        stage_chunks = calculate_stage_chunks(read_chunks, write_chunks, stage_count)
+        pre_chunks = [read_chunks] + stage_chunks
+        post_chunks = stage_chunks + [write_chunks]
+
+        int_chunks = [
+            _calculate_shared_chunks(pre, post)
+            for pre, post in zip(pre_chunks, post_chunks)
+        ]
+        if all(itemsize * prod(chunks) >= min_mem for chunks in int_chunks):
+            # success!
+            return list(zip(pre_chunks, int_chunks, post_chunks))
```

Comment:
Another option would be to terminate the loop when the estimated number of read/write operations stops decreasing. This might not achieve the `min_mem` target, but would indicate the point at which there is nothing further to be gained by adding stages. This can actually happens pretty quickly, e.g., in my ERA5 example the number of IO ops starts increases beyond 7 stages.